convlstm post (now ready for review) (#133)
rstudio/ai-blog: _posts/2020-12-17-torch-convlstm/torch-convlstm.Rmd

---
title: "Convolutional LSTM for spatial forecasting"
description: > 

>   In forecasting spatially-determined phenomena (the weather, say, or the next frame in a movie), we want to model temporal evolution, ideally using recurrence relations, while efficiently extracting spatial features, as usually done using convolutions. Ideally, we'd have at our disposal an architecture that is both recurrent and convolutional. In this post, we build a convolutional LSTM with torch.

Author: backticks don't work in the abstract ... but otherwise ... yeah ... unfortunately all that is correct :-(
will see what I can do

author:
  - name: Sigrid Keydana
    affiliation: RStudio
    affiliation_url: https://www.rstudio.com/
slug: keydanatorchconvlstm
date: 12-17-2020
categories:
  - Torch
  - R
  - Image Recognition & Image Processing
  - Time series
output:
  distill::distill_article:
    self_contained: false
    toc: true
preview: images/preview.jpeg
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

This post is the first in a loose series aimed at exploring forecasting of spatially-determined data over time. By spatially-determined I mean that whatever the quantities we're trying to predict -- be they univariate or multivariate time series, of spatial dimensionality or not -- the *input* data are given on a spatial grid.

> For example, the input could be atmospheric measurements, such as sea surface temperature or pressure, given at some set of latitudes and longitudes. The target to be predicted could then span that same (or another) grid. Alternatively, it could be a univariate time series, like a meteorological index.

Author: yes :-)

But wait a second, you may be thinking. For time series prediction, we have that time-honored set of recurrent architectures, right -- LSTM, GRU ...? Right. We do; but once we feed spatial data to an RNN, treating different locations as different input features, we lose an essential structural relationship. Importantly, we need to operate in both space and time. We want both: recurrence relations and convolutional filters. Enter *convolutional RNNs*.

## What to expect from this post

> Today, we won't jump into real-world applications just yet. Instead, we'll take our time to build a convolutional LSTM (henceforth: convLSTM) in `torch`. For one, we have to -- there is no official PyTorch implementation. (As opposed to Keras, by the way. If you're interested in quickly playing around with a Keras convLSTM, check out [this nice example](https://tensorflow.rstudio.com/guide/keras/examples/conv_lstm/).)

Author: great idea!

## A `torch` convLSTM

The code discussed here may be found on [Github](https://github.com/skeydan/convlstm). (Depending on when you're reading this, the code in that repository may have evolved though.)

My starting point was one of the PyTorch implementations found on the net, namely, [this one](https://github.com/ndrplz/ConvLSTM_pytorch/blob/master/convlstm.py). If you search for "PyTorch convGRU" or "PyTorch convLSTM", you will find stunning discrepancies in how these are realized -- discrepancies not just in syntax and/or engineering ambition, but on the semantic level, right at the center of what the architectures may be expected to do. As they say, let the buyer beware. (Regarding the realization I ended up porting, I am confident that while numerous optimizations will be possible, the basic mechanism matches my expectations.

What do I expect? Let's approach this task in a top-down way.

### Input and output

The convLSTM's input will be a time series of spatial data, each observation being of size `(time steps, channels, height, width)`.

> Compare this with the usual RNN input format, be it in `torch` or Keras. In both frameworks, RNNs expect tensors of size `(timesteps, input_dim)`[^1]. `input_dim` is $1$ for univariate time series and greater than $1$ for multivariate ones. Conceptually, we may match this to convLSTM's `channels` dimension: There could be a single channel, for temperature, say -- or there could be several, such as for pressure, temperature, and humidity. The two additional dimensions found in convLSTM, `height` and `width`, are spatial indexes into the data.

Author: lol :-) I tried again ;-) ... I really prefer `indexes` (but that may be because I had several years (((working as a DBA))) to digest it ;-)))

Re technical language

> Meanwhile, indices is generally preferred in mathematical, financial, and technical contexts, while indexes is relatively common in general usage.

I'm not sure how they mean it, but personally I'd only ever use indices when referring to i, j, k or the like, but otherwise ... like in the above sentence ... I think indexes sounds more natural ...

However, with RNNs, it is not all about outputs. RNN architectures, in addition, carry through hidden states.

What are hidden states? I carefully phrased that sentence to be as general as possible -- deliberately circling around the confusion that, in my view, often arises at this point. We'll attempt to clear up some of that confusion in a second, but let's first finish our high-level requirements specification.

We want our convLSTM to be usable in different contexts and applications. Various architectures exist that make use of hidden states, most prominently perhaps, encoder-decoder architectures. Thus, we want our convLSTM to return those as well. Again, this is something a `torch` LSTM does by default, while in Keras it is achieved using `return_state = TRUE`.

Now though, it really is time for that interlude. We'll sort out the ways things are called by both `torch` and Keras, and inspect what you get back from their respective GRUs and LSTMs.

### Interlude: Outputs, states, hidden values ... what's what?

> For this to remain an interlude, I intend to summarize findings on a high level. The code snippets in the appendix show how to arrive at these results. Heavily commented, they probe return values from both Keras and `torch` GRUs and LSTMs. Running these will make the upcoming summaries seem a lot less abstract.

Author: lol :-) don't force me to say ;-)

[^2]: In theory, it would be possible for them to be of different sizes if the respective weight matrices transformed their operands to the same output size.

$$
\mathbf{h}_t = \mathbf{W}_{x}\mathbf{x}_t + \mathbf{W}_{h}\mathbf{h}_{t-1}
$$

Now, *about those hidden states*.

When a Keras LSTM is defined with `return_state = TRUE`, its return value is a structure of three entities called output, memory state, and carry state. In `torch`, the same entities are referred to as output, hidden state, and cell state. (In `torch`, we always get all of them.)

So are we dealing with three different types of entities? We are not.

The cell, or carry, state is that special thing setting apart LSTMs from GRUs deemed responsible for the "long" in "long short-term memory". Technically, it could be reported to the user at all points in time; as we'll see shortly though, it is not.

> What about outputs and hidden, or memory, states? Confusingly, these really are the same thing. Recall that for each input *item* in the input *sequence*, we're combining it with the previous state, resulting in a new state, to be made used of in the next step[^3]:

Author: interesting - I didn't know!

+------------------------------------------------------------------------------------------------------------------------------------+---------------+--------------+

: Table 1: RNN terminology. Comparing torch-speak and Keras-speak. In row 1, the terms are parameter names. In rows 2 and 3, they are pulled from current documentation.
:::

<br />

Now, about that public vs. private distinction. In both frameworks, we can obtain outputs (hidden states) for every time step. The cell state, however, we can access only for the very last time step. This, it need be said, is purely an implementation decision. As we'll see when building our own recurrent module, there are no obstacles inherent in keeping track of cell states and passing them back to the user.

If you dislike the pragmatism of this distinction, you can always go with the math. When a new cell state has been computed (based on prior cell state, input, forget, and cell gates -- the specifics of which we are not going to get into here), it is transformed to the hidden (a.k.a. output) state making use of yet another, the output gate:

$$
h_t = o_t \odot \tanh(c_t)
$$

> Definitely, then, hidden state (output, resp.) builds on cell state, adding additional modeling power. Whether one would interpret the workings of the output gate in an anthropomorphizing way is left to personal choice.

Reviewer: This is probably because I don't understand the technical aspects, but I'm not even sure how one would go about interpreting the output gate as though it were human. If you think this last sentence adds something, then you can leave it. But, given the amount of prose in this piece (all prior to actually building the convLSTM), I'd consider cutting it.

Author: you're right as always - this is a leftover hinting at an -  even longer - paragraph that got completely removed (because verbosity ... yeah ....)

| access all intermediate outputs and the final "cell state"                        | both of the above | `return_sequences = TRUE, return_state = TRUE` |
+-----------------------------------------------------------------------------------+-------------------+------------------------------------------------+
| access all intermediate outputs and "cell states" from all time steps             | no way            | no way                                         |
+-----------------------------------------------------------------------------------+-------------------+------------------------------------------------+

: Table 2: Contrasting ways of obtaining various return values in `torch` vs. Keras. Cf. the appendix for complete examples.
:::

<br />

### `convLSTM`, the plan

In both `torch` and Keras RNN architectures, single time steps are processed by corresponding `Cell` classes: There is an LSTM Cell matching the LSTM, a GRU Cell matching the GRU, and so on. We do the same for ConvLSTM. In `convlstm_cell()`, we first define what should happen to a single observation; then in `convlstm()`, we build up the recurrence logic.

> Once we're done, we create a dummy dataset, as reduced-to-the-essential as can be. With more complex datasets, even artificial ones, chances are that if we don't see any training progress, there are hundreds of possible explanations. We want a sanity check that, if we fail it, leaves no excuses. Realistic applications are left to future posts.

Reviewer: ```suggestion
Once we're done, we create a dummy dataset, as reduced-to-the-essentials as can be. With more complex datasets, even artificial ones, chances are that if we don't see any training progress, there are hundreds of possible explanations. We want a sanity check that, if failed, leaves no excuses. Realistic applications are left to future posts.
```

### A single step: `convlstm_cell`

Our `convlstm_cell`'s constructor takes arguments `input_dim` , `hidden_dim`, and `bias`, just like a `torch` LSTM Cell.

> But we're processing two-dimensional input data. Instead of the usual affine combination of new input and previous state, we use a convolution, of kernel size `kernel_size`. Inside `convlstm_cell`, it is `self$conv` that takes care of this.

Reviewer: ```suggestion
But we're processing two-dimensional input data. Instead of the usual affine combination of new input and previous state, we use a convolution of kernel size `kernel_size`. Inside `convlstm_cell`, it is `self$conv` that takes care of this.
```

> Note how the `channels` dimension, which in the original input data would correspond to different variables, is creatively used to consolidate four convolutions into one: Each channel output will be passed to just one of the four cell gates. Once in possession of the convolution output, `forward()` applies the gate logic, resulting in the two types of states it needs to send back to the caller.

Reviewer: This paragraph is ⭐️!

Author: hehe thanks :-))

```{r}
dim(layer_outputs[[1]])
# [1]  2  4  5 16 16
```

`layer_last_states[[1]]`is a list of tensors, the first of which holds the single layer's final hidden state, and the second, its final cell state.

```{r}
dim(layer_last_states[[1]][[1]])
# [1]  2  5 16 16
dim(layer_last_states[[1]][[2]])
# [1]  2  5 16 16
```

For comparison, this is how return values look for a multi-layer architecture.

```{r}
model <- convlstm(input_dim = 3, hidden_dims = c(5, 5, 1), kernel_sizes = rep(3, 3), n_layers = 3)
c(layer_outputs, layer_last_states) %<-% model(x)

# for each layer, tensor of size (batch_size, seq_len, hidden_size, height, width)
dim(layer_outputs[[1]])
# 2  4  5 16 16
dim(layer_outputs[[3]])
# 2  4  1 16 16

# list of 2 tensors for each layer
length(layer_last_states)
# 3

# h, of size (batch_size, hidden_size, height, width)
dim(layer_last_states[[3]][[1]])
# 2  1 16 16

# c, of size (batch_size, hidden_size, height, width)
dim(layer_last_states[[3]][[2]])
# 2  1 16 16
```

Now we want to sanity-check this module with the simplest-possible dummy data.

### Sanity-checking the `convlstm`

We generate black-and-white "movies" of diagonal beams successively translated in space.

> Each sequence consists of six time steps, and each beam, of six pixels. Just a single sequence is created manually. To create that one sequence, we start from a single beam:

Reviewer: ```suggestion
Each sequence consists of six time steps, and each beam of six pixels. Just a single sequence is created manually. To create that one sequence, we start from a single beam:
```

```{r}
round(as.matrix(preds[1, 1, 6:15, 10:19]), 2)
```

           [,1]  [,2]  [,3]  [,4]  [,5]  [,6]  [,7]  [,8]  [,9] [,10]
     [1,]  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00     0
     [2,] -0.02  0.36  0.01  0.06  0.00  0.00  0.00  0.00  0.00     0
     [3,]  0.00 -0.01  0.71  0.01  0.06  0.00  0.00  0.00  0.00     0
     [4,] -0.01  0.04  0.00  0.75  0.01  0.06  0.00  0.00  0.00     0
     [5,]  0.00 -0.01 -0.01 -0.01  0.75  0.01  0.06  0.00  0.00     0
     [6,]  0.00  0.01  0.00 -0.07 -0.01  0.75  0.01  0.06  0.00     0
     [7,]  0.00  0.01 -0.01 -0.01 -0.07 -0.01  0.75  0.01  0.06     0
     [8,]  0.00  0.00  0.01  0.00  0.00 -0.01  0.00  0.71  0.00     0
     [9,]  0.00  0.00  0.00  0.01  0.01  0.00  0.03 -0.01  0.37     0
    [10,]  0.00  0.00  0.00  0.00  0.00  0.00 -0.01 -0.01 -0.01     0

This should suffice for a sanity check. If you made it till the end, thanks for your patience! In the best case, you'll be able to apply this architecture (or a similar one) to your own data -- but even if not, I hope you've enjoyed learning about `torch` model coding and/or RNN weirdness ;-)

> Myself, I am certainly looking forward to exploring convLSTMs on real-world problems in the near future. Thanks for reading!

Reviewer: Myself is a reflexive pronoun, which you only use when it's the object of the sentence. The best way I can think of it is with french verbs, _se laver_ to wash (you wash yourself). You can put yourself to bed, bathe yourself, kill yourself…there aren't a lot of reflexive verbs in English (or at least I can't think of them right now). Basically, people use myself incorrectly all the time (including above). Myself can also be used as an intensive pronoun, but, in that case, it should go after the "I"— _I myself am certainly looking forward to…_

c(layer_outputs, layer_last_states) %<-% model(x)
```

We get back a list of length two, which we immediately split up into the two types of output returned: intermediate outputs from all layers, and final states (of both types) for the last layer.

With just a single layer, `layer_outputs[[1]]`holds all of the layer's intermediate outputs, stacked on dimension two.

```{r}
dim(layer_outputs[[1]])
# [1]  2  4  5 16 16
```

`layer_last_states[[1]]`is a list of tensors, the first of which holds the single layer's final hidden state, and the second, its final cell state.

```{r}
dim(layer_last_states[[1]][[1]])
# [1]  2  5 16 16
dim(layer_last_states[[1]][[2]])
# [1]  2  5 16 16
```

For comparison, this is how return values look for a multi-layer architecture.

```{r}
model <- convlstm(input_dim = 3, hidden_dims = c(5, 5, 1), kernel_sizes = rep(3, 3), n_layers = 3)
c(layer_outputs, layer_last_states) %<-% model(x)

# for each layer, tensor of size (batch_size, seq_len, hidden_size, height, width)
dim(layer_outputs[[1]])
# 2  4  5 16 16
dim(layer_outputs[[3]])
# 2  4  1 16 16

# list of 2 tensors for each layer
length(layer_last_states)

> # 3

Reviewer 2: It took me some time to understand the comment and the length here. Maybe print the `str` instead?

Author: great idea, done!

beam <- torch_eye(6) %>% nnf_pad(c(6, 12, 12, 6)) # left, right, top, bottom
beams[[1]] <- beam
```

Using `torch_roll()` , we create a pattern where this beam moves up diagonally, and stack the individual tensors along the `timesteps` dimension.

```{r}
for (i in 2:6) {
  beams[[i]] <- torch_roll(beam, c(-(i-1),i-1), c(1, 2))
}

init_sequence <- torch_stack(beams, dim = 1)
```

That's a single sequence. Thanks to `torchvision::transform_random_affine()`, we almost effortlessly produce a dataset of a hundred sequences. Moving beams start at random points in the spatial frame, but they all share that upward-diagonal motion.

```{r}
sequences <- vector(mode = "list", length = 100)
sequences[[1]] <- init_sequence

for (i in 2:100) {
  sequences[[i]] <- transform_random_affine(init_sequence, degrees = 0, translate = c(0.5, 0.5))
}

input <- torch_stack(sequences, dim = 1)
dim(input)
# [1] 100   6  24  24
```

That's it for the raw data. Now we still need a `dataset` and a `dataloader`. Of the six time steps, we use the first five as input and try to predict the last one.

```{r}
dummy_ds <- dataset(
  
  initialize = function(data) {
    # add channels dimension

>     self$data <- data$unsqueeze(3)

Reviewer 2: i found it confusing for this to be there. I would make the `input` sequence with the correct size isntead.

Author: makes sense!

  recurrent_initializer = initializer_constant(value = 1)
)
gru(input)

# return_state = TRUE, return_sequences = TRUE
# return shape = list of:
#    - outputs, of shape: (batch_size, time steps, units)
#    - "memory states" for the last time step, of shape: (batch_size, units)
#
# note how again, the "memory state" found in list item 2 matches the final-time-step outputs reported in item 1
gru <- layer_gru(
  units = 1,
  return_sequences = TRUE,
  return_state = TRUE,
  kernel_initializer = initializer_constant(value = 1),
  recurrent_initializer = initializer_constant(value = 1)
)
gru(input)
```

### `torch`

#### LSTM (non-stacked architecture)

```{r}
library(torch)

# batch of 3, with 4 time steps each and a single feature
# we will specify batch_first = TRUE when creating the LSTM
input <- torch_randn(c(3, 4, 1))
input

# default args
# return shape = (batch_size, units)
#
# note: there is an additional argument num_layers that we could use to specify a stacked LSTM - effectively composing two LSTM modules
# default for num_layers is 1 though 
lstm <- nn_lstm(
  input_size = 1, # number of input features
  hidden_size = 1, # number of hidden (and output!) features

>   batch_first = TRUE # for easy comparability with Keras

Reviewer 2: This is not really really equivalent to keras, because keras has `bias=FALSE` by default. I know this doesn't change the point, so maybe just ignore that :)

Author: oh, but Keras has `use_bias = TRUE` it seems?
